### PR TITLE
[PROPOSAL] increase qa test reliability

### PR DESCRIPTION
**Pull request by gandrewstone**
In 2 ways:

1. sync_mempools has never worked reliably because there is a bloom filter in front of INV relay and bloom filters are probabilistic.  This change forces INVs to be pushed between nodes randomly if the mempool is not syncing.

2. Either error message can be returned when a block's transaction validation fails.

### CODE REVIEW

File: qa/rpc-tests/test_framework/util.py [hunk 1]
```diff
@@ -287,6 +291,18 @@ def sync_mempools(rpc_connections, wait=1,verbose=1):
             logging.info("sync mempool: " + str(pool_len))
         if num_match == len(rpc_connections):
             break
+
+        if ((count&3)==0):
+            # mempools may never sync fully because there is a bloom filter that can block some INVs
+            # so randomly push all txs from one node to another until the sync happens
```
Comment (ptschip) [inline qa/rpc-tests/test_framework/util.py:301]: Not sure I really like the idea of this section. What I'm afraid of is that we'll end up masking issues with tx propagation, such as with the fast filter and an issue could slip past us into mainnet at some point.  The inv false positive should be 1 in a million, at least that's the false positive rate of the old style rolling bloom filters, so it doesn't seem like something we need to fix.

Comment (gandrewstone) [reply qa/rpc-tests/test_framework/util.py:301]: We have timeouts and tests interrupted due to ctrl-c all the time.  Maybe it's not travis fault.  Suspiciously, these problems happen often in wallet backup.py, sendrawtx.py and others which whenever I look have mempool sync.  

I think the problem is that the chance isn't 1 in a million if the bloom is set for that.  It's the birthday paradox.  And this it just for 1 call.  How many sync calls per test run?  And then travis has 9 platforms it runs on.

We need the mempool sync to actually work 100% reliably so we dont waste so much time on weird failures, and test the filter performance separately.  I have a test for fast filter collision but right now it's set much higher than 1 in a million.

But I still need to work on this.  It doesn't solve the filters in AlreadyReceivedTx so mempool sync still isn't 100%.

File: qa/rpc-tests/walletbackup.py [hunk 2]
```diff
@@ -136,7 +136,7 @@ def run_test(self):
 
         # Generate 101 more blocks, so any fees paid mature
```
Comment (ptschip) [inline qa/rpc-tests/walletbackup.py:137]: I think there also needs to be a sync at the end of do_one_round:

#Must sync mempools before mining
self.nodes[3].generate(1)
+self.sync_blocks()

File: qa/rpc-tests/test_framework/util.py (hunk 1 above)
Comment (ptschip) [reply qa/rpc-tests/test_framework/util.py:301]: I think this one I can't agree with.  Not until we've moved away from Travis to GitLab and see where the chips fall there.  As far as I've seen , yes we've always had some spurious issue in wallet and wallet_backup but mostly the tests run well with the exception lately of Travis timing out the build process.

Comment (gandrewstone) [reply qa/rpc-tests/test_framework/util.py:301]: Ok. Fair enough.  Prior to the parallel mempool admission PR, I fixed a bunch of reliability issues in the qa tests so we really haven't had the chance to run with those fixes and without the k=1 fastfilter (that has more collisions).  

Right now if I run the full suite on my machine a bunch of times and leave, I'm sure to be hung with unsynchronized mempools when I get back.  So I'm going to finish this PR up as a "demonstrator" (there's not much left to do) but let's not merge it.  

I'll then fix fastfilter and let's see how well things work.

**Comment by ptschip** [comment]
ok, but it would be good to get at least the bug for the miner_tests.cpp merged in since that one is failing everybodys tests right now the most.

File: src/test/miner_tests.cpp [hunk 3]
```diff
@@ -330,7 +349,7 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity)
     hash = tx.GetHash();
     mempool.addUnchecked(hash, entry.Fee(1000000).Time(GetTime()).SpendsCoinbase(false).FromTx(tx));
     BOOST_CHECK_EXCEPTION(BlockAssembler(chainparams_regtest).CreateNewBlock(scriptPubKey), std::runtime_error,
-        HasReason("bad-blk-signatures"));
+        HasEitherReason("mandatory-script-verify-flag-failed", "bad-blk-signatures"));
```
Comment (ptschip) [inline src/test/miner_tests.cpp:352]: see #1351

File: qa/rpc-tests/walletbackup.py (hunk 2 above)
Comment (gandrewstone) [reply qa/rpc-tests/walletbackup.py:137]: merged your PR

**Pull request closed without merging**